src/objective/hinge.cu

```diff
@@ -38,6 +38,8 @@ class HingeObj : public ObjFunction {
     const bool is_null_weight = info.weights_.Size() == 0;
     const size_t ndata = preds.Size();
     out_gpair->Resize(ndata);
+    GPUSet devices = (preds.Devices().IsEmpty() && tparam_->external_memory)
```

Comment:
You can make a policy for this configuration, and abstract the policy into a method of `GPUSet` or another class.  But again I think it's not the best heuristic.

Reply:
@trivialfis - agree that this is not a better heuristic. i'll add another configuration item to control this.